File: golang/backspace_string_compare.go
package main

/*
Space	: O(n)
Time	: O(n)

method:
Stack
*/

func processBackspace(S string) string {
	a := ""
	ls := len(S)
	for i := 0; i < ls; i++ {
		if string(S[i]) == "#" {
			if len(a) > 0 {
				a = a[:len(a)-1]
			}
		} else {
			a += string(S[i])
		}
	}
	return a
}

func backspaceCompare(S string, T string) bool {
	return processBackspace(S) == processBackspace(T)
}

File: python3/backspace_string_compare.py
"""
Space	: O(n)
Time	: O(n)

Mehtod:
Stack
"""


class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        def process(s):
            a = []
            n = len(s)
            for i in range(n):
                if s[i] == '#':
                    if a:
                        a.pop()
                else:
                    a.append(s[i])
            return a

        return process(s) == process(t)
